Stripe Connect integration

allouis opened the pull request:
no-issue

This is a WIP of wiring up Ghost to work with Stripe Connect and includes a few changes/additions

- `stripe-connect` module in the members service
	- This handles creating the stripe connect auth link & parsing the result
- `membersStripeConnect` controller
    - This is an API controller which issues a redirect to the stripe auth link
- `settings` controller
    - This has been updated to accept `stripe_connect_integration_tokens` in order to configure Stripe Connect
- `MembersConfigProvider` The existing members config has been wrapped in a class, so that we can pass settings/config to it, making it easier and quicker to test

@naz I would _love_ if you could look at the changes in the API, both the `membersStripeConnect` and `settings` controllers, check it makes sense to you, and that I'm not breaking any patterns!

@ErisDS It would be cool for you to look at the last commit, and check the `stripeDirect` config switch is working as you wanted it to (I've tried to give the tests decent names)

allouis committed:
Refactored the settings edit controller

no-issue

There was some unused code here, the variable  was never used, also we
were looping and collecting a list of errors, but only every using the
first one, so switched to the `find` method which stops iteration after
an element has matched.

allouis committed:
Added stripe connect module to member service

no-issue

This module handles the creation of a url used for authorization of
Stripe Connect, and also the parsing of the data eventually received
from the authorization flow.

allouis committed:
Added stripe_connect_integration default setting

no-issue

This will be used to store the keys and other information after a
sucessful Stripe Connect OAuth flow.

allouis committed:
Wired up stripe-connect module to settings API

no-issue

Uses the members service to parse a stripe_connect_integration_token
setting and set the stripe_connect_integration based on that.

This change includes ignoring the stripe_connect_integration{,_token}
settings, as the token is never saved, and the integration can only be
set by using the token.

allouis committed:
Added membersStripeConnect controller auth method

no-issue

In order to issue a redirect we need access to the "raw" req/res
objects, which is why we must return the function which gets access to
them.

The members service is used to create the auth url and to update the
users session.

allouis committed:
Refactored members config to use DI

no-issue

This makes testing it much easier

allouis committed:
Updated members config to respect stripeDirect

no-issue

If the stripeDirect config value is NOT set / false, then connect token
should be used over the API keys if both are present, or else use
whichever is present.  If the stripeDirect config value is set /
true,then API keys should be used and stripe connect token should be
completely ignored.

The reason for this is that we want to use Stripe Connect as the primary
auth method going forward, but we want to keep the direct version
availiable should the Ghost Foundation ever dismantle. This will allow
people to use all the same features with no dependency on an external
service.

allouis commented on core/server/api/canary/membersStripeConnect.js:
```diff
@@ -0,0 +1,17 @@
+
+module.exports = {
+    auth: {
+        permissions: false,
```
Needs to be `true`

ErisDS commented:
Commit message appears to state the opposite of the intended logic:

> > This will never use the connect keys if stripeDirect is set to false, if
it is set to true, it will prefer the connect keys, but use whatever
exists

The logic (slightly updated with explicir false/true) is meant to be:

- If the `stripeDirect` config value is NOT set / false(normal case), then connect token should be used over the API keys if both are present, or else use whichever is present.
- If the `stripeDirect` config value is set / true, then API keys should be used and stripe connect token should be completely ignored.

Tests appear to do the right thing though? I'm confused 🙈 

I suggest the commit message should be the explicit intended logic and also (as always) include a why - this config value exists for 2 reasons: futureproofing and development.

allouis commented:
:grimacing: Will fix the commit - that's the bit that's wrong!

allouis committed:
Added permission for auth:members_stripe_connect

no-issue

This permission is used to ensure that only the Owner can
authenticate/authorize the members_stripe_connect integration.

allouis commented:
@naz @daniellockyer @kevinansfield Would you be able to check out the migration in [https://github.com/TryGhost/Ghost/pull/11861/commits/677eb018358a2b029a025cc0fcb14b76557364aa ](https://github.com/TryGhost/Ghost/pull/11861/commits/677eb018358a2b029a025cc0fcb14b76557364aa) please?

naz commented on test/unit/models/settings_spec.js:
```diff
@@ -113,8 +114,12 @@ describe('Unit: models/settings', function () {
+                    }, []).length;
                     // 2 events per item - settings.added and settings.[name].added
-                    eventSpy.callCount.should.equal(92);
+                    eventSpy.callCount.should.equal(numberOfSettings * 2);
```
:heart:

allouis commented on core/server/api/canary/settings.js:
```diff
@@ -124,6 +131,21 @@ module.exports = {
                 }
             }
 
+            if (stripeConnectIntegrationToken) {
```
I can't see how the current stripe config is protected by permissions, so it seems like this should work the same way it currently does - @naz do you have any insight on this?

naz replied:
@allouis there is no specific permission checks for member related config values. The only thing that is protected is toggling the "members" flag itself - https://github.com/TryGhost/Ghost/commit/786eaac57ecdbfcbc4c7969f3e04c1023b537927 That said, it's not what we'd want ideally. 

What I see happening is having specific permissions once stripe configuration is extracted into separate endpoint - then it would be a lot more manageable "resource-based" approach to implement. Think current approach is OK as is, but ideally would become more granular in the future.

allouis replied:
Okay cool - that makes total sense, thank! :relaxed:

daniellockyer commented on core/server/data/migrations/versions/3.18/001-add-members_stripe_connect-auth-permissions.js:
```diff
@@ -0,0 +1,56 @@
+const ObjectId = require('bson-objectid');
+const {logging} = require('../../../../lib/common');
```
logging is now in `core/shared/logging` :eyes:

allouis replied:
Fixed in latest commit [https://github.com/TryGhost/Ghost/pull/11861/commits/a13c471b572735b1755715f3925dab4c365a877e ](https://github.com/TryGhost/Ghost/pull/11861/commits/a13c471b572735b1755715f3925dab4c365a877e)

naz commented on core/server/web/api/canary/admin/routes.js:
```diff
@@ -97,6 +97,8 @@ module.exports = function apiRoutes() {
         http(apiCanary.members.importCSV)
     );
 
+    router.get('/members/stripe-connect', shared.middlewares.labs.members, mw.authAdminApi, http(apiCanary.membersStripeConnect.auth));
```
Not sure about the naming for `stripe-connect`. We don't have the `-` convention for any other endpoints and usually use `_` as a separator, e.g.: `/email_preview/posts/:id`, `/integrations/:id/api_key/:keyid/refresh`, `/members/:id/signin_urls`

allouis replied:
Cool - will raise w/ Hannah/Zimo RE naming :+1:

allouis replied:
Updated to use underscore, I think bothering them with naming q like that is waste of time!

naz commented on core/server/api/canary/membersStripeConnect.js:
```diff
@@ -0,0 +1,18 @@
+        permissions: true,
+        query(frame) {
+            // This is something you have to do if you want to use the "framework" with access to the raw req/res
+            frame.response = async function (req, res) {
```
This is a tough one tbh. This enpoint has completely different concept to all the other ones as it's not "resource-based". Instead, it performs an action - redirect. We only use redirects on middleware level so far and having it here feels very strange. That said, don't have a better concept on where we could put this piece :thinking: 

Does the API has to do the redirect to the `stripeConnectAuthURL` or can the client do the redirect based on the response from the server?

allouis replied:
> This is a tough one tbh. This enpoint has completely different concept to all the other ones as it's not "resource-based".

Yeah I think our API framework is quite restrictive in that sense, it makes (IMO) valid usecases quite difficult if they deviate from the norm!


> That said, don't have a better concept on where we could put this piece 

Yeah I could think of anywhere better either!

> Does the API has to do the redirect to the stripeConnectAuthURL or can the client do the redirect based on the response from the server?

Technically it could be done on the frontend, a discussion was had, and this was the simpler approach - IMO restructuring so we generate/redirect-to the URL on the client is not worth it though! Would be nicer if controllers could have easier ability to do things like this

naz replied:
> Technically it could be done on the frontend, a discussion was had, and this was the simpler approach

After sleeping on it, here is my thinking. If it's not absolutely impairing for the client experience,  we should follow returning "redirect_to_url" property on the returned resource approach. First, tt would be consistent with the resource-oriented convention we try to follow. Second, wont create any hacky solutions which don't fit anywhere in the "frame". 

A good reference for this could be how the redirect URL is generated by Stripe in case of creating a checkout page or [setup intent](https://stripe.com/docs/api/setup_intents/object#setup_intent_object-next_action-redirect_to_url) with `redirect_to_url` property.

I'm probably lacking a lot of context here, so treat this comment purely from a perspective of "how I'd do it the clean way" :sweat_smile: Lmk your thoughts!

allouis replied:
That solution seems convoluted to me, they'd click a link and we'd make a request, parse the response and then go to some location! That's what the `Location` header is for, and it means being able to use a simple `<a>` tag on the frontend!

I get that you want it to be cleaner, but I really think this needs to be fixed at the "frame" level. We shouldn't be restricted to only supporting certain types of requests because of that, right?

naz replied:
Cool. The `<a>` tag usecase makes sense  :+1: Let's allow it as is then and see if we have any similar usecases in the future to modify the "frame"

naz commented on core/server/data/migrations/versions/3.18/001-add-members_stripe_connect-auth-permissions.js:
```diff
@@ -0,0 +1,56 @@
+            action_type: 'auth',
+            object_type: 'members_stripe_connect',
+            created_at: date,
+            created_by: 1,
```
Think using hardcoded `1` here is not the safest approach for couple reasons: 1) one can transfer "owner" role to other user 2) one can remove the user after the transfer which would make this piece break (no user with `id:1`).
Noticed this logic is used in [another migration](https://github.com/TryGhost/Ghost/blob/935acfdb88824f24e04a13b8a0ca647066ba2a36/core/server/data/migrations/versions/3.12/01-add-identity-permission.js#L31) as well. Suggest we fix this. Same goes for `updated_by` column.

allouis replied:
This is what would be used if set by the model layer https://github.com/TryGhost/Ghost/blob/d2cf7111ec3a77a0f6454783a43ea79ddb6da3ae/core/server/models/base/index.js#L661

Fixing this is way outside the scope of this issue IMO

naz replied:
@allouis there is very little code that relies on the "internalUser" and `1` as hardcoded value. There was a push towards eliminating use of it in this [issue](https://github.com/TryGhost/Ghost/issues/9299).  I think adding these hard-coded values in migrations adds to a problem which hasn't been fully eliminated. Would rather go with additional query to figure out what is the current "owner id". Alternatively, would add this piece of code to be tracked in the issue above, so once it's on the table again we know where to clean up. 

One thing to bear in mind, having these `1`s in migrations "cements" them here for long time as we can't retroactively fix them.

naz replied:
This issue is also on the [roadmap](https://github.com/TryGhost/Team/issues/221) for Ghost 4.0 under "Review v0.1 related "dirty hotfix" in contextUser method"

allouis replied:
>  I think adding these hard-coded values in migrations adds to a problem which hasn't been fully eliminated.

From what I can tell, literally every migration we currently have uses the id of `1` for the {created,updated}_by fields.

Until we have a plan of how it should be handled I don't think we should start changing things now and deviating from the norm - what do you think?

naz replied:
Yup, agreed. Forgot about `context:internal` using the `internalUser` id in the end of the day :disappointed: Let's roll with the `1` :+1:

naz commented:
@allouis, checked the migration all works well on both MySQL and SQLite. Left a comment around hard-coded id which think we'd need to handle properly. The rest looks :ok_hand:

kevinansfield commented on core/server/data/migrations/versions/3.18/02-add-members_stripe_connect-auth-permissions.js:
```diff
@@ -0,0 +1,56 @@
+
+        const date = connection.raw('CURRENT_TIMESTAMP');
+
+        await connection('permissions').insert({
+            id: ObjectId.generate(),
+            name: 'Auth Stripe Connect for Members',
+            action_type: 'auth',
+            object_type: 'members_stripe_connect',
+            created_at: date,
+            created_by: 1,
+            updated_at: date,
+            updated_by: 1
+        });
```
Is there a reason this isn't using the existing permissions fixtures?

allouis replied:
Yes! Because they go via the model layer, and I don't think we should be using the model layer for migrations anymore!

kevinansfield replied:
👍 OK. Would be good to make a bit of noise about this so that everyone is aware we want to change patterns in migrations

kevinansfield reviewed (approved):
Tested migration, worked for me and code looks 👍
